theano/compile/builders.py

```diff
@@ -115,40 +173,197 @@ def __hash__(self):
         # TODO: use internal variables in hash
         return hash(type(self))
 
-    def make_node(self, *inputs):
-        for input, type in zip(inputs, self.input_types):
-            if not type == input.type:
-                raise TypeError("Wrong type, expected %s but got %s" %
-                                (type, input.type))
-        return gof.Apply(self,
-                         list(inputs) + self.shared_inputs,
-                         [type() for type in self.output_types])
+    def _recompute_grad_op(self):
+        if isinstance(self._grad_op, OpFromGraph):
+            self._grad_op_is_cached = True
+            return
+        output_grads = [out_t() for out_t in self.output_types]
+        if self._grad_op is None:
+            self._grad_op = []
+
+        # we need to convert a list/function into an OfG instance
+        if isinstance(self._grad_op, list):
+            goverrides_l = self._grad_op
+            if len(goverrides_l) > len(self.local_inputs):
+                raise ValueError(
+                    'Can override %d gradients at most, got %d' % (
+                        len(self.local_inputs), len(goverrides_l)),
+                    self.goverrides_l)
+            if len(goverrides_l) < len(self.local_inputs):
+                goverrides_l += [None] * (
+                    len(self.local_inputs) - len(goverrides_l))
+            wrt_l = [lin for lin, gov in
+                     izip(self.local_inputs, goverrides_l) if not gov]
+            # compute non-overriding downsteam grads from upstreams grads
+            # it's normal some input may be disconnected, thus the 'ignore'
+            gdefaults = iter(theano.gradient.grad(
+                cost=None,
+                known_grads=dict(izip(self.local_outputs, output_grads)),
+                wrt=wrt_l,
+                disconnected_inputs='ignore') if wrt_l else [])
+            # combine overriding gradients
+            all_grads_l = []
+            for inp, gov in izip(self.local_inputs, goverrides_l):
+                if gov is None:
+                    all_grads_l.append(next(gdefaults))
+                elif gov is undef:
+                    all_grads_l.append(
+                        inp.zeros_like().astype(theano.config.floatX))
```

Comment:
Alternatively, this can be `int(0) -> zeros_like()` and `None -> None`, this preserves literal meaning of `None`. Just a interface matter, it's your call @nouiz.